api/src/services/export/dhis.js

```diff
@@ -0,0 +1,215 @@
+ * @param {Object[]} contacts The set of contact documents relevant to the calculation
+ * @returns {Object} The @param contacts _id values mapped to an array of orgUnits above them in the hierarchy
+ */
+const mapContactIdToOrgUnits = (dataSet, contacts, contactsWithOrgUnits) => {
+  const result = {};
+  for (const contact of contactsWithOrgUnits) {
+    const dhisConfigs = Array.isArray(contact.dhis) ? contact.dhis : [contact.dhis];
+    for (const dhisConfig of dhisConfigs) {
+      const dataSetMatch = !dhisConfig.dataSet || dhisConfig.dataSet === dataSet;
+      if (dhisConfig.orgUnit && dataSetMatch) {
+        if (!result[contact._id]) {
+          result[contact._id] = [];
+        }
+
+        result[contact._id].push(dhisConfig.orgUnit);
+      }
+    }
+  }
+  
+  for (const contact of contacts) {
+    let traverse = contact;
+    while (traverse) {
+      if (result[traverse._id]) {
+        const existing = result[contact._id] || [];
+        result[contact._id] = _.uniq([...existing, ...result[traverse._id]]);
+      }
+
+      traverse = traverse.parent;
+    }
+  }
+
+  return result;
+};
+
+const buildDataValues = (targetDefinitions, targetDocs, orgUnits) => {
+  const mapTargetIdToDhis = ObjectFromEntries(targetDefinitions.map(target => ([target.id, target.dhis])));
+  const createEmptyValueSetFor = orgUnit => {
+    const result = {};
+    for (const target of targetDefinitions) {
+      const { dataElement } = target.dhis;
+      result[dataElement] = Object.assign(
+        {},
+
+        /*
+        Copies any attribute defined in dhis config onto the dataValue
+        Primary usecase is `categoryOptionCombo` and `attributeOptionCombo` but there are many others
+        */
+        target.dhis,
+        { orgUnit, value: 0 },
+      );
+      delete result[dataElement].dataSet;
+    }
+    return result;
+  };
+
+  // all results start with 0s
+  const dataValueSet = {};
+  for (const contactOrgUnits of Object.values(orgUnits)) {
+    for (const orgUnit of contactOrgUnits) {
+      if (!dataValueSet[orgUnit]) {
+        dataValueSet[orgUnit] = createEmptyValueSetFor(orgUnit);
+      }
+    }
+  }
+
+  // add relevant values onto the result
+  for (const targetDoc of targetDocs) {
+    const unitsOfOwner = orgUnits[targetDoc.owner];
+    if (!unitsOfOwner) {
+      continue;
+    }
+
+    for (const orgUnit of unitsOfOwner) {
+      for (const target of targetDoc.targets) {
+        const dataElement = mapTargetIdToDhis[target.id] && mapTargetIdToDhis[target.id].dataElement;
+        if (dataElement) {
+          const dataValueObj = dataValueSet[orgUnit][dataElement];
+          if (dataValueObj) {
+            dataValueObj.value += target.value.total;
+          }
+        }
+      }
+    }
+  }
+
+  return _.flatten(Object.values(dataValueSet).map(dataValueGroup => Object.values(dataValueGroup)));
+};
+
+const makeHumanReadable = (response, dataSetConfig, dhisTargetDefinitions, contacts) => {
```

Comment:
A brief comment explaining the goal of this would be nice (what is changing to make it human readable)

Reply:
Also, are we actually exposing this feature anywhere? It's not used by the UI correct? So this would be just something that people who directly called the API would be able to use

Reply:
It isn't currently exposed anywhere, no. But the primary interface to this will actually be an app that runs inside DHIS2 that will be built by a tech-lead.